Improve the Unified Resources card UI

bl-nero opened the pull request:
- Allow extending and collapsing if labels don't fit in the container.
- Make labels clickable.

[Demo](https://www.loom.com/share/fedaf22026134a3bbb4c60a5ef2d9bd2?sid=37c33a0c-654c-4f43-a27b-59e204cbbe85)

Related issue: #27462

avatus committed:
Update resourceKind method to return slice

avatus committed:
Adds a UnifiedResource watcher to the auth server. This watcher
will watch all the types that are displayed in the web UI and store
them in-memory to allow us to search/filter/query and get multiple
kinds returned at the same time.
[RFD](https://github.com/gravitational/teleport/pull/28162)

avatus committed:
Add UnifiedResourceComponent

avatus committed:
Add unifiedResourceWatcher to auth server

avatus committed:
Init unified resource watcher

avatus committed:
Convert to strings for error output

avatus committed:
Add license

bl-nero committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Add some tests

bl-nero committed:
Merge remote-tracking branch 'origin/master' into michaelmyers/unified_resource_watcher

bl-nero committed:
Fix a race condition and properly close the watcher

avatus committed:
add defer

avatus committed:
Add kubes to watcher

avatus committed:
Add uninit check

avatus committed:
Notify if stale

avatus committed:
Add SAML to watcher

avatus committed:
Update test and fix lint

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Fix license

avatus committed:
Add godocs

avatus committed:
Add Kind to ui structs

avatus committed:
Add Kind to ui structs

avatus committed:
Add FakePaginateParams

avatus committed:
Add FakePaginateParams

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update ListUnifiedResources proto messages

avatus committed:
Update uses of FakePaginate

avatus committed:
Update uses of FakePaginate

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Include type in seenmap for MatchResourceByFilters

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
Update TestAuthServer with unified resource watcher

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
CheckAndSetDefaults for ListUnifiedResourcesRequest

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Add ListUnifiedResources rpc

avatus committed:
Update protos

avatus committed:
Update protos

avatus committed:
Add test

avatus committed:
Add test

avatus committed:
Add docs and update tests

avatus committed:
Add docs and update tests

avatus committed:
Add more docs

avatus committed:
Add more docs

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Combine app types in switch

avatus committed:
Combine app types in switch

avatus committed:
Fix tests

avatus committed:
Fix tests

avatus committed:
Add kube access checker

avatus committed:
Add kube access checker

avatus committed:
Add remaining kinds

avatus committed:
Add remaining kinds

bl-nero committed:
Add a Unified Resource card component

bl-nero committed:
WIP: Unified Resource list

bl-nero committed:
Lint fixes

avatus committed:
Review feedback

avatus committed:
Add configurable queuesize to NewWatcher

avatus committed:
Move Get to collector

bl-nero committed:
Review comments

bl-nero committed:
Rename AgentKind and fix lint issues

avatus committed:
make unified resource cache public

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

avatus committed:
Move to common interface

avatus committed:
Use copy as value

avatus committed:
Move saml getter

bl-nero committed:
Review comments

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
WIP: infinite scrolling

bl-nero committed:
Prevent VS Code from showing errors in tsconfig

If tsconfig doesn't have an `outDir` option set, VS Code shows multitude of errors of the following kind:

```
Cannot write file '/Users/bartosz/code/teleport/e/web/teleport/babel.config.js' because it would overwrite input file.
```

bl-nero committed:
Remove unnecessary files

bl-nero committed:
todo

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Cleanup

bl-nero committed:
Add license

avatus committed:
invert watcher and cache

avatus committed:
use cache instead of watcher

avatus committed:
Fix imports

avatus committed:
Fix lint: check for err

avatus committed:
review round 1

avatus committed:
Update test

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Review, fixing broken tests

avatus committed:
Add copy to get

bl-nero committed:
Add a search panel to the unified resources list

avatus committed:
Add CloneAny interface to help with unified resource cloning

avatus committed:
Use cloneany

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Add sort for name and kind

avatus committed:
Add app and saml, fix tests

bl-nero committed:
Review

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Review

bl-nero committed:
Lint

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

avatus committed:
Watch kube server instead of cluster

avatus committed:
Watch kube server instead of cluster

avatus committed:
Merge branch 'master' into michaelmyers/unified_resource_watcher

avatus committed:
Get kubecluster from server

avatus committed:
merge master

avatus committed:
update proto file

avatus committed:
Add no access test case

bl-nero committed:
Review, fix CI problems

bl-nero committed:
One more lint fix

bl-nero committed:
Yet another lint fix

avatus committed:
move sorting

avatus committed:
Fix sorting by name

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

avatus committed:
Review feedback round 2

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Remove unneeded bk return

avatus committed:
Merge branch 'michaelmyers/unified_resource_watcher' into michaelmyers/unified_grpc

avatus committed:
Keep grpc naming convention

avatus committed:
Add missing docs

avatus committed:
defer log

bl-nero committed:
WIP: filter and sort

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Make sorting work

bl-nero committed:
Make filters (almost) work

bl-nero committed:
Fix the filters

bl-nero committed:
Fix filter history handling and kind-based sorting

bl-nero committed:
Merge branch 'michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Visual tweaks

bl-nero committed:
lint and test fixes

bl-nero committed:
Add licenses

bl-nero committed:
WIP: show more labels

bl-nero committed:
Improve the card UI, remove checkboxes, make labels clickable

bl-nero committed:
Clean up, add comments

bl-nero committed:
Review

bl-nero committed:
Merge branch 'bl-nero/unified-resources-sort' into bl-nero/unified-resources-labels

avatus commented on web/packages/teleport/src/UnifiedResources/ResourceCard.tsx:
```diff
@@ -54,60 +52,146 @@ const ResTypeIconBox = styled(Box)`
+                const { name, value } = label;
+                const labelText = `${name}: ${value}`;
+                return (
+                  <ResourceLabel
+                    key={labelText}
+                    title={labelText}
+                    onClick={() => onLabelClick?.(label)}
+                    kind="secondary"
+                    data-is-label=""
+                  >
+                    {labelText}
+                  </ResourceLabel>
+                );
```
We've ran into issues before with duplicate labels causing some weird bugs that would infinitely spawn more and more labels as the user used the UI. [check it out](https://github.com/gravitational/teleport.e/pull/722).
Generally speaking, duplicate labels _shouldn't_ happen but they can as we've seen so we need to add an extra identifier.

bl-nero replied:
Hmm, the change that you mentioned is about duplicate _resources_, not labels, isn't it? The labels are just controls that are used for display.

Also, that PR contains a very weird corner case: if I had resources called `foo1` on the 5th position in the list, and a resource called `foo` on the 15th position, they'd share the same key. :)

In the hindsight, I should probably also use something like `JSON.stringify([name, value])` for dodging all possible corner cases, since I still may run into a bug if someone puts a colon in the label name or value.

avatus commented on web/packages/teleport/src/UnifiedResources/Resources.tsx:
```diff
@@ -77,7 +84,7 @@ export function Resources() {
       <ResourcesContainer gap={2}>
         {fetchedData.agents.map((agent, i) => (
-          <ResourceCard key={i} resource={agent} />
+          <ResourceCard key={i} resource={agent} onLabelClick={onLabelClick} />
```
Because these resources will change order a lot (with sorting/filtering), we can't use the index as the key. Since `UnifiedResource` is a type union between all our agents, we'll need to settle on a common "key" that they all share in their types, uniquely. Unforunately, that doesn't exist current. Half of them have a `name` and the other half have an `id`. I'd be fine with just adding one of those to the missing agents and crafting it however we need to, avoiding duplication by maybe appending `kind` to it? something like `name/kind` or `id/kind`.


We can also think about doing this in the backend if it makes more sense there. You let me know

bl-nero replied:
You're right, that's not how it should be done. That's the reason why I launched an email thread once about unifying resource IDs, but I think we didn't reach any conclusion. It can be a UUID, a `kind/name` pair, or whatever; from the client side perspective, it should be an opaque string, and we shouldn't even attempt to parse it. All the information that could potentially be derived, like the resource kind, should be passed in a structured form anyway; this way we allow ourselves some degree of freedom and we won't need to hunt all the pieces of code that manipulate strings on the client side to potentially fetch useful information (and half the time doing it wrong). And yes, it means it should be the server's responsibility to generate this. I'll take care of it in a separate PR, if you don't mind.

bl-nero replied:
Added an issue for this: https://github.com/gravitational/teleport/issues/30259

avatus reviewed (commented):
Styles seem great on my end, let a few comments

bl-nero committed:
Fix the useEffect dependency list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Merge branch 'bl-nero/unified-resources-sort' into bl-nero/unified-resources-labels

avatus committed:
Review feedback

ravicious commented on web/packages/teleport/src/UnifiedResources/ResourceCard.tsx:
```diff
@@ -54,60 +52,146 @@ const ResTypeIconBox = styled(Box)`
+        // components. If we don't do it, there's no way to sync the background
+        // of the "more" button, since it's opaque and the background is
+        // animated.
+        className="grv-unified-resource-card"
```
We should be able to get away with not using class names directly here, but it's fine for me if we leave it like this. In that case, maybe let's add a TODO comment to address this in the future.

https://stackoverflow.com/a/41010688/742872

ravicious reviewed (approved):
Leaving a preemptive approval in case that I'm away and you want to merge this ASAP.

The implementation looks and works fine.

bl-nero commented on web/packages/teleport/src/UnifiedResources/ResourceCard.tsx:
```diff
@@ -54,60 +52,146 @@ const ResTypeIconBox = styled(Box)`
+  // This effect installs a resize observer whose purpose is to detect the size
+  // of the component that contains all the labels. If this component is taller
+  // than the height of a single label row, we show a "+x more" button.
+  React.useEffect(() => {
```
Wow, thanks!

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_grpc' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Merge branch 'bl-nero/unified-resources-sort' into bl-nero/unified-resources-labels

bl-nero committed:
Address part of the review comments

avatus committed:
Add ListUnifiedResources gRPC and web endpoints (#29661)

Adds the `ListUnifiedResources` grpc endpoint that returns paginated unified resources and exposes it in the web apiserver

bl-nero committed:
Merge remote-tracking branch 'origin/michaelmyers/unified_resource_watcher' into bl-nero/unified-resources-card

bl-nero committed:
Merge branch 'bl-nero/unified-resources-card' into bl-nero/unified-resources-list

bl-nero committed:
Merge branch 'bl-nero/unified-resources-list' into bl-nero/unified-resources-search

bl-nero committed:
Merge branch 'bl-nero/unified-resources-search' into bl-nero/unified-resources-sort

bl-nero committed:
Merge branch 'bl-nero/unified-resources-sort' into bl-nero/unified-resources-labels

bl-nero committed:
Merge branch 'master' into bl-nero/unified-resources-labels

bl-nero committed:
Merge branch 'master' into bl-nero/unified-resources-labels
